Sources/Core/ObjectiveCInitExtension.swift

```diff
@@ -66,26 +76,64 @@ extension ObjCModelRenderer {
         }
     }
 
-    public func renderInitWithModelDictionary() -> ObjCIR.Method {
+    func expectedObjectType(schema: Schema) -> String? {
+        switch schema {
+        case .object,
+             .map(valueType: _):
+            return "NSDictionary"
+        case .array(itemType: _),
+             .set(itemType: _):
+            return "NSArray"
+        case .integer,
+             .float,
+             .boolean,
+             .enumT(.integer):
+            return "NSNumber"
+        case .string(format: _),
+             .enumT(.string):
+            return "NSString"
+        default:
+            return nil
+        }
+    }
+
+    public func renderInitWithModelDictionaryError() -> ObjCIR.Method {
         func renderPropertyInit(
             _ propertyToAssign: String,
             _ rawObjectName: String,
+            keyPath: [String],
             schema: Schema,
             firstName: String, // TODO: HACK to get enums to work (not clean)
-            counter: Int = 0
+            counter: Int = 0,
+            dirtyPropertyName: String? = nil,
+            needsTypeCheck: Bool = true
         ) -> [String] {
+            func renderTypeCheck(_ body: () -> [String]) -> [String] {
+                guard needsTypeCheck else { return body() }
+                guard let kindOf = expectedObjectType(schema: schema) else { return body() }
+                let key = keyPath.count == 1 ? keyPath[0] : "[@[\(keyPath.joined(separator: ", "))] componentsJoinedByString:@\".\"]"
+                return [
+                    ObjCIR.ifStmt("!error || [\(rawObjectName) isKindOfClass:[\(kindOf) class]]", body: body) +
```

Comment:
I think this is the source and should be `&&`?

Reply:
No, `||` is what I intended here, but it's not super obvious because it's a bit of a logical optimization.

The condition is "we have an error object and want type checking" or "we don't have an error object so skip type checking".